### DISCUSSION TITLE
stop loss = 0 with long orders is throwing a ValueError

### DISCUSSION THREAD

**Discussion by shparekh** [kernc/backtesting.py#1078] [Q&A] [answered]
Hello,
I can't seem to figure out why I get a ValueError when stop loss = 0.   I do not intend to use limit orders.

I have following next() in my strategy  - 
```
    def next(self):
        super().init()
        price = self.data.Close
        # Long Exit
        if (self.position) and crossover(self.lema, self.lrs):
            # Close any existing long trades, and sell the asset
            self.position.close()
        # Long Entry
        elif (not self.position) and crossover(self.lrs, self.lema) and (price >= self.ema):
            self.buy(sl=(1-self.stop_loss/100)*price)

```
I am running it with -
```
    bt = Backtest(df, myStrategy, cash=10000, commission=0, trade_on_close=True, exclusive_orders=False, hedging=False)
```
with stop_loss = 0, it throws following error - 

```
File ~\AppData\Roaming\Python\Python311\site-packages\backtesting\backtesting.py:713, in _Broker.new_order(self, size, limit, stop, sl, tp, trade)
    711 if is_long:
    712     if not (sl or -np.inf) < (limit or stop or adjusted_price) < (tp or np.inf):
--> 713         raise ValueError(
    714             "Long orders require: "
    715             f"SL ({sl}) < LIMIT ({limit or stop or adjusted_price}) < TP ({tp})")
    716 else:
    717     if not (tp or -np.inf) < (limit or stop or adjusted_price) < (sl or np.inf):

ValueError: Long orders require: SL (0.4557679295539856) < LIMIT (0.4557679295539856) < TP (None)
```

What am I doing wrong?

Thank you.
Best regards

**Comment by initial-d** [comment, marked as answer]
The error is expected when `stop_loss = 0` with this expression:

```python
self.buy(sl=(1 - self.stop_loss / 100) * price)
```

If `stop_loss` is zero, then `sl == price`. For a long order, Backtesting.py requires the stop-loss to be below the entry/limit price, not equal to it. So the validation fails because the order is effectively saying "enter long and stop out at the same price".

If you do not want to use a stop-loss, pass `sl=None` instead of calculating a zero-distance stop:

```python
price = self.data.Close[-1]

sl = None if self.stop_loss == 0 else (1 - self.stop_loss / 100) * price

self.buy(sl=sl)
```

A couple of small code notes:

- Use `self.data.Close[-1]` for the current bar's close, not `self.data.Close`, which is the data series wrapper.
- You should not call `super().init()` inside `next()`. If you need `super().init()`, it belongs in your strategy's `init()` method.
- If you are using `trade_on_close=True`, be careful that an entry based on the current close and a stop calculated from that same close can be very tight. A tiny stop percentage may still fail or behave unexpectedly if the required order relation is not `SL < entry < TP` for a long trade.

So the short version is: `0` should mean "no stop" in your strategy logic, but the library interprets the number you pass to `sl` as an actual stop price. Use `None` when stop-loss is disabled.
